[T153220354] Fix header inclusions in c10

q10 opened the pull request:
Summary: - Break up D45769819 to fix fix header inclusions in c10.

This is a re-attempt to land the iwyu header changes, by taking the diff from [PR 100304](https://github.com/pytorch/pytorch/pull/100304), and adding the bare minimal changes to make the diff build corectly in the internal builds.

Test Plan:
```
buck2 build --no-remote-cache mode/dev-nosan //caffe2/c10/...

buck2 build --no-remote-cache mode/dev-nosan //deeplearning/fbgemm/fbgemm_gpu/...
```

Differential Revision: D45920611



cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @mcarilli @ptrblck @leslie-fang-intel @cyyever

albanD reviewed (changes requested):
Could you please add a detailed description of what you're doing here?
Did you use the iwyu config we have in the repo to generate this?

q10 commented:
> Could you please add a detailed description of what you're doing here? Did you use the iwyu config we have in the repo to generate this?

Hi @albanD, I have updated the description with details on the change.  It is essentially [PR 100304](https://github.com/pytorch/pytorch/pull/100304) with the bare minimum additions required to fix internal builds, so yes, I am effectively using the iwyu config.

q10 committed:
[T153220354] Fix header inclusions in c10 (#1541)

Summary:
This is a re-attempt to land the iwyu header changes, by taking the diff from [PR 100304](https://github.com/pytorch/pytorch/pull/100304), and adding the bare minimal changes to make the diff build corectly in the internal builds.

X-link: https://github.com/facebookresearch/pytorch3d/pull/1541

X-link: https://github.com/fairinternal/pytorch3d/pull/44

Pull Request resolved: https://github.com/pytorch/pytorch/pull/101623

- Re-work D45769819 to fix header inclusions in c10

Test Plan:
```
buck2 build --no-remote-cache mode/dev-nosan //caffe2/c10/...

buck2 build --no-remote-cache mode/dev-nosan //deeplearning/fbgemm/fbgemm_gpu/...

buck2 build mode/dev-nosan //vision/fair/pytorch3d/pytorch3d:_C
```

Reviewed By: malfet

Differential Revision: D45920611

fbshipit-source-id: 6c694e45a7b4ca50a316f1662ad684ad753aa24a
